panic: bug: ClusterConfig called on closed or nonexistent connection

I have had two different Syncthing instances crash on Windows 10. One is running v0.14.28, one v.0.14.29-rc1.
Both are using SyncTrayzor (V1.1.16).

I restarted one, only to have it crash ~5 mins later. I captured the Synctrazor log from that one and have attached it.

I will endeavour to provide any other information required.

Forum post for reference: [https://forum.syncthing.net/t/syncthing-crashing/9845](url)

[SyncTrayzor.txt](https://github.com/syncthing/syncthing/files/1021902/SyncTrayzor.txt)

For clarity:

```
[5NSF2] 09:17:06 INFO: Failed to exchange Hello messages with K2ECF46 (95.97.92.221:22067): read tcp 192.168.1.50:61924->95.97.92.221:22067: i/o timeout 
entWatcher: Received 0 events 
panic: bug: ClusterConfig called on closed or nonexistent connection 
goroutine 4242 [running]: 
github.com/syncthing/syncthing/lib/model.(*Model).ClusterConfig(0xc042058280, 0xdbf211def3228856, 0xc92a31f0bf8e42e1, 0xd87e29914adce15e, 0x87221d62cccbc3d9, 0xc042631f00, 0x1, 0x1) 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/model/model.go:789 +0x21c3 
github.com/syncthing/syncthing/lib/protocol.(*nativeModel).ClusterConfig(0xc042b91fe0, 0xdbf211def3228856, 0xc92a31f0bf8e42e1, 0xd87e29914adce15e, 0x87221d62cccbc3d9, 0xc042631f00, 0x1, 0x1) 
	<autogenerated>:74 +0x95 
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).readerLoop(0xc0457315c0, 0x0, 0x0) 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:306 +0x11de 
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Start 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:171 +0x4a 
goroutine 1 [chan receive, 70 minutes]: 
main.syncthingMain(0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0xc042057a40, 0x34, 0x0, 0x0, ...) 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:954 +0x1c5a 
main.main() 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:459 +0x541 
goroutine 20 [select, 71 minutes]: 
github.com/syncthing/syncthing/vendor/github.com/xtaci/kcp-go.(*updateHeap).updateTask(0x10d0f40) 
```

--- a/lib/protocol/common_test.go
+++ b/lib/protocol/common_test.go
@@ -13,6 +13,7 @@ type TestModel struct {
 	hash          []byte
 	weakHash      uint32
 	fromTemporary bool
+	indexFn       func(DeviceID, string, []FileInfo)
 	closedCh      chan struct{}
 	closedErr     error
 }
@@ -24,6 +25,9 @@ func newTestModel() *TestModel {
 }
 
 func (t *TestModel) Index(deviceID DeviceID, folder string, files []FileInfo) {
+	if t.indexFn != nil {
+		t.indexFn(deviceID, folder, files)
+	}
 }
 
 func (t *TestModel) IndexUpdate(deviceID DeviceID, folder string, files []FileInfo) {

--- a/lib/protocol/protocol.go
+++ b/lib/protocol/protocol.go
@@ -182,11 +182,14 @@ type rawConnection struct {
 	nextID    int32
 	nextIDMut sync.Mutex
 
-	outbox        chan asyncMessage
-	closed        chan struct{}
-	closeOnce     sync.Once
-	sendCloseOnce sync.Once
-	compression   Compression
+	inbox               chan message
+	outbox              chan asyncMessage
+	clusterConfigBox    chan *ClusterConfig
+	receiverLoopStopped chan struct{}
+	closed              chan struct{}
+	closeOnce           sync.Once
+	sendCloseOnce       sync.Once
+	compression         Compression
 }
 
 type asyncResult struct {
@@ -220,15 +223,18 @@ func NewConnection(deviceID DeviceID, reader io.Reader, writer io.Writer, receiv
 	cw := &countingWriter{Writer: writer}
 
 	c := rawConnection{
-		id:          deviceID,
-		name:        name,
-		receiver:    nativeModel{receiver},
-		cr:          cr,
-		cw:          cw,
-		awaiting:    make(map[int32]chan asyncResult),
-		outbox:      make(chan asyncMessage),
-		closed:      make(chan struct{}),
-		compression: compress,
+		id:                  deviceID,
+		name:                name,
+		receiver:            nativeModel{receiver},
+		cr:                  cr,
+		cw:                  cw,
+		awaiting:            make(map[int32]chan asyncResult),
+		inbox:               make(chan message),
+		outbox:              make(chan asyncMessage),
+		clusterConfigBox:    make(chan *ClusterConfig),
+		receiverLoopStopped: make(chan struct{}),
+		closed:              make(chan struct{}),
+		compression:         compress,
 	}
 
 	return wireFormatConnection{&c}
@@ -237,8 +243,9 @@ func NewConnection(deviceID DeviceID, reader io.Reader, writer io.Writer, receiv
 // Start creates the goroutines for sending and receiving of messages. It must
 // be called exactly once after creating a connection.
 func (c *rawConnection) Start() {
+	go c.readerLoop()
 	go func() {
-		err := c.readerLoop()
+		err := c.receiverLoop()
 		c.internalClose(err)
 	}()
 	go c.writerLoop()
@@ -348,25 +355,37 @@ func (c *rawConnection) ping() bool {
 	return c.send(&Ping{}, nil)
 }
 
-func (c *rawConnection) readerLoop() (err error) {
+func (c *rawConnection) readerLoop() {
 	fourByteBuf := make([]byte, 4)
+	for {
+		msg, err := c.readMessage(fourByteBuf)
+		if err != nil {
+			if err == errUnknownMessage {
+				// Unknown message types are skipped, for future extensibility.
+				continue
+			}
+			c.internalClose(err)
+			return
+		}
+		select {
+		case c.inbox <- msg:
+		case <-c.closed:
+			return
+		}
+
+	}
+}
+
+func (c *rawConnection) receiverLoop() (err error) {
+	defer close(c.receiverLoopStopped)
+	var msg message
 	state := stateInitial
 	for {
 		select {
+		case msg = <-c.inbox:
 		case <-c.closed:
 			return ErrClosed
-		default:
 		}
-
-		msg, err := c.readMessage(fourByteBuf)
-		if err == errUnknownMessage {
-			// Unknown message types are skipped, for future extensibility.
-			continue
-		}
-		if err != nil {
-			return err
-		}
-
 		switch msg := msg.(type) {
 		case *ClusterConfig:
 			l.Debugln("read ClusterConfig message")
@@ -846,6 +865,8 @@ func (c *rawConnection) internalClose(err error) {
 			}
 		}
 		c.awaitingMut.Unlock()
+
+		<-c.receiverLoopStopped
 
 		c.receiver.Closed(c, err)
 	})

--- a/lib/protocol/protocol.go
+++ b/lib/protocol/protocol.go
@@ -182,14 +182,13 @@ type rawConnection struct {
 	nextID    int32
 	nextIDMut sync.Mutex
 
-	inbox               chan message
-	outbox              chan asyncMessage
-	clusterConfigBox    chan *ClusterConfig
-	receiverLoopStopped chan struct{}
-	closed              chan struct{}
-	closeOnce           sync.Once
-	sendCloseOnce       sync.Once
-	compression         Compression
+	inbox                 chan message
+	outbox                chan asyncMessage
+	dispatcherLoopStopped chan struct{}
+	closed                chan struct{}
+	closeOnce             sync.Once
+	sendCloseOnce         sync.Once
+	compression           Compression
 }
 
 type asyncResult struct {
@@ -223,18 +222,17 @@ func NewConnection(deviceID DeviceID, reader io.Reader, writer io.Writer, receiv
 	cw := &countingWriter{Writer: writer}
 
 	c := rawConnection{
-		id:                  deviceID,
-		name:                name,
-		receiver:            nativeModel{receiver},
-		cr:                  cr,
-		cw:                  cw,
-		awaiting:            make(map[int32]chan asyncResult),
-		inbox:               make(chan message),
-		outbox:              make(chan asyncMessage),
-		clusterConfigBox:    make(chan *ClusterConfig),
-		receiverLoopStopped: make(chan struct{}),
-		closed:              make(chan struct{}),
-		compression:         compress,
+		id:                    deviceID,
+		name:                  name,
+		receiver:              nativeModel{receiver},
+		cr:                    cr,
+		cw:                    cw,
+		awaiting:              make(map[int32]chan asyncResult),
+		inbox:                 make(chan message),
+		outbox:                make(chan asyncMessage),
+		dispatcherLoopStopped: make(chan struct{}),
+		closed:                make(chan struct{}),
+		compression:           compress,
 	}
 
 	return wireFormatConnection{&c}
@@ -245,7 +243,7 @@ func NewConnection(deviceID DeviceID, reader io.Reader, writer io.Writer, receiv
 func (c *rawConnection) Start() {
 	go c.readerLoop()
 	go func() {
-		err := c.receiverLoop()
+		err := c.dispatcherLoop()
 		c.internalClose(err)
 	}()
 	go c.writerLoop()
@@ -376,8 +374,8 @@ func (c *rawConnection) readerLoop() {
 	}
 }
 
-func (c *rawConnection) receiverLoop() (err error) {
-	defer close(c.receiverLoopStopped)
+func (c *rawConnection) dispatcherLoop() (err error) {
+	defer close(c.dispatcherLoopStopped)
 	var msg message
 	state := stateInitial
 	for {
@@ -866,7 +864,7 @@ func (c *rawConnection) internalClose(err error) {
 		}
 		c.awaitingMut.Unlock()
 
-		<-c.receiverLoopStopped
+		<-c.dispatcherLoopStopped
 
 		c.receiver.Closed(c, err)
 	})

--- a/lib/protocol/protocol_test.go
+++ b/lib/protocol/protocol_test.go
@@ -125,6 +125,53 @@ func TestCloseOnBlockingSend(t *testing.T) {
 	}
 }
 
+func TestCloseRace(t *testing.T) {
+	indexReceived := make(chan struct{})
+	unblockIndex := make(chan struct{})
+	m0 := newTestModel()
+	m0.indexFn = func(_ DeviceID, _ string, _ []FileInfo) {
+		close(indexReceived)
+		<-unblockIndex
+	}
+	m1 := newTestModel()
+
+	ar, aw := io.Pipe()
+	br, bw := io.Pipe()
+
+	c0 := NewConnection(c0ID, ar, bw, m0, "c0", CompressNever).(wireFormatConnection).Connection.(*rawConnection)
+	c0.Start()
+	c1 := NewConnection(c1ID, br, aw, m1, "c1", CompressNever)
+	c1.Start()
+	c0.ClusterConfig(ClusterConfig{})
+	c1.ClusterConfig(ClusterConfig{})
+
+	c1.Index("default", nil)
+	select {
+	case <-indexReceived:
+	case <-time.After(time.Second):
+		t.Fatal("timed out before receiving index")
+	}
+
+	go c0.internalClose(errManual)
+	select {
+	case <-c0.closed:
+	case <-time.After(time.Second):
+		t.Fatal("timed out before c0.closed was closed")
+	}
+
+	select {
+	case <-m0.closedCh:
+		t.Errorf("receiver.Closed called before receiver.Index")
+	default:
+	}
+
+	close(unblockIndex)
+
+	if err := m0.closedError(); err != errManual {
+		t.Fatal("Connection should be closed")
+	}
+}
+
 func TestMarshalIndexMessage(t *testing.T) {
 	if testing.Short() {
 		quickCfg.MaxCount = 10

